internal/cluster/discovery.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/hazelcast/hazelcast-go-client/cluster/discovery"
+	"github.com/hazelcast/hazelcast-go-client/hzerrors"
+	ilogger "github.com/hazelcast/hazelcast-go-client/internal/logger"
+)
+
+const (
+	discoveryStrategyAdapterStarted = 1
+)
+
+type DiscoveryStrategyAdapter struct {
+	strategy    discovery.Strategy
+	lg          ilogger.LogAdaptor
+	translator  map[pubcluster.Address]pubcluster.Address
+	mu          *sync.RWMutex
+	opts        discovery.StrategyOptions
+	started     int32
+	usePublicIP bool
+}
+
+func NewDiscoveryStrategyAdapter(cfg pubcluster.DiscoveryConfig, lg ilogger.LogAdaptor) *DiscoveryStrategyAdapter {
+	if cfg.Strategy == nil {
+		panic("discovery strategy is required")
+	}
+	return &DiscoveryStrategyAdapter{
+		strategy:    cfg.Strategy,
+		translator:  map[pubcluster.Address]pubcluster.Address{},
+		usePublicIP: cfg.UsePublicIP,
+		mu:          &sync.RWMutex{},
+		lg:          lg,
+		opts: discovery.StrategyOptions{
+			Logger:      lg,
+			UsePublicIP: cfg.UsePublicIP,
+		},
+	}
+}
+
+func (d *DiscoveryStrategyAdapter) Addresses(ctx context.Context) ([]pubcluster.Address, error) {
+	d.lg.Trace(func() string {
+		return "cluster.DiscoveryStrategyAdapter.Addresses"
+	})
+	if err := d.ensureStarted(ctx); err != nil {
+		return nil, err
+	}
+	nodes, err := d.strategy.DiscoverNodes(ctx)
+	if err != nil {
+		return nil, err
+	}
+	res := make([]pubcluster.Address, 0, len(nodes))
+	for _, node := range nodes {
+		if node.PrivateAddr == "" && node.PublicAddr == "" {
+			continue
+		}
+		if node.PrivateAddr == "" {
+			res = append(res, (pubcluster.Address)(node.PublicAddr))
+			continue
+		}
+		if node.PublicAddr == "" {
+			res = append(res, (pubcluster.Address)(node.PrivateAddr))
+			continue
+		}
+		if d.usePublicIP {
+			res = append(res, (pubcluster.Address)(node.PublicAddr))
+			continue
+		}
+		res = append(res, (pubcluster.Address)(node.PrivateAddr))
+	}
+	return res, nil
+}
+
+func (d *DiscoveryStrategyAdapter) TranslateMember(ctx context.Context, member *pubcluster.MemberInfo) (addr pubcluster.Address, err error) {
+	d.lg.Trace(func() string {
+		return fmt.Sprintf("cluster.DiscoveryStrategyAdapter.TranslateMember: %s %s", member.UUID.String(), member.Address)
+	})
+	if err = d.ensureStarted(ctx); err != nil {
+		return
+	}
+	addr, ok := d.translate(member.Address)
+	if !ok {
+		nodes, err := d.strategy.DiscoverNodes(ctx)
+		if err != nil {
+			return addr, err
+		}
+		d.updateTranslation(nodes)
+		addr, ok = d.translate(member.Address)
+	}
+	if !ok {
+		return addr, hzerrors.ErrAddressNotFound
+	}
+	return addr, nil
+}
+
+func (d *DiscoveryStrategyAdapter) ensureStarted(ctx context.Context) error {
+	if atomic.LoadInt32(&d.started) != discoveryStrategyAdapterStarted {
+		// discovery strategy was not started yet, try to start it
+		if err := d.start(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *DiscoveryStrategyAdapter) start(ctx context.Context) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
```

Comment:
Don't we have this lock for the 

```go
// run strategy's Start method if it implements that.
	if starter, ok := d.strategy.(discovery.StrategyStarter); ok {
		d.lg.Debug(func() string {
			return "Starting discovery strategy"
		})
		if err := starter.Start(ctx, d.opts); err != nil {
			return err
		}
	}
```

block? Does moving lock and unlock to the borders of this block make sense? 🤔

Reply:
I didn't understand the question. whole `(d *DiscoveryStrategyAdapter) start` function is protected with the lock.

Reply:
I am questioning if we need to protect the whole method, won't it be enough to protect only the critical area? Also will doing it improve performance?

Reply:
The only parts in that function except the block is getting/setting the value of `d.started` which should be done under the lock, that doesn't have any affect on the performance.